### DISCUSSION TITLE
Removing underscore from _id

### DISCUSSION THREAD

**Discussion by vresetnikov** [typegoose/typegoose#676] [Q&A]
Hello,

Sorry if my question is very simplistic, but being still new to mongodb, typescript and typegoose I am struggling to find information on how to remove the underscore ( _ ) from _id in typegoose. 

I found the following example for the normal mongoose: 

```javascript
schema.method("toJSON", function() {
    const { __v, _id, ...object } = this.toObject();
    object.id = _id;
    return object; 
```

however, seeing how typegoose takes care of schemas for me, I can't seem to figure out how to achieve the desired goal of removing the underscore from the returned id. 

Thank you for your precious time

**Comment by smolinari** [comment]
@vresetnikov - Mongoose has a virtual for the `_id` field called `id` on all returned documents. Just add the property `id: string` to your entity. :grin: 

https://mongoosejs.com/docs/api.html#document_Document-id

Scott

**Reply by smolinari** [reply]
This is an example of code using the `id` property.
```ts
export class TodoItemEntity implements Base {
  _id!: Types.ObjectId;

  id!: string;

  @Prop({ required: true })
  title!: string;

  @Prop()
  description?: string;
 ```
  Scott

**Reply by vresetnikov** [reply]
Hello, thank you for your reply. 

Based on your response I simply did as follows:


```typescript
export class DatesTemplate {

    @prop({default: () => nanoid()})
    id!: string;

    @prop({required: true, unique: true})
    templateName: string;

    @prop({ required: true, type: () => datesObject })
    dates!: datesObject[];

}
```

Could you kindly let me know if this is stable/ correct? By the way, why the exclamation mark after id in this case?

Thank you so much!

**Reply by hasezoey** [reply]
> Could you kindly let me know if this is stable/ correct?

no this is not correct, if you just want a type (for something like a plugin provides or mongoose itself provides), then you should not set a decorator on the property (in this case, remove `@prop({default: () => nanoid()})`)
if you still want custom id's, then you need to modify `_id` with a decorator, `id` is just a virtual that is basically an alias for `_id` that converts to string

> By the way, why the exclamation mark after id in this case?

this basically tells typescript `this property always exists, dont question it`, which is required if you use `strict: true`, or just want to be explicit, the opposite is `?`, which basically tells typescript `this property can be undefined, dont question it`

a sign(`?` or `!`) is required in typescript with typegoose and tsconfig `strict: true`, because typescript would otherwise expect these properties to be set in the constructor, which typegoose classes dont use / have

**Reply by vresetnikov** [reply]
@hasezoey thanks for your reply. After having read your comment I have unsuccessfully attempted to apply your suggestion:

```Typescript

export interface dateMap {
    date: Date,
    isActive: boolean,
    description: string,
}

class datesObject implements dateMap {

    @prop({default: () => nanoid()})
    _id!: mongoose.Types.ObjectId;

    @prop()  //no effect if prop decorator is removed.
    id!: string;

    @prop({type: () => Date, unique: true, required: true})
    public date: Date;

    @prop({type: () => Boolean, required: true})
    public isActive: boolean;

    @prop({type: () => String, required: true})
    public description: string;

}

export class DatesTemplate {

    @prop({default: () => nanoid()})
    _id!: mongoose.Types.ObjectId;

    @prop() //no effect with the prop decorator removed
    id!: string;

    @prop({required: true, unique: true})
    templateName!: string;

    @prop({ required: true, type: () => datesObject })
    dates!: datesObject[];

}

```
However, the id without the underscore is still not appearing, nor in the main, nor within the children. 

What am I doing wrong?

**Reply by hasezoey** [reply]
> However, the id without the underscore is still not appearing, nor in the main, nor within the children.

where and how are you testing this? also, what versions are you using? (typescript, mongoose, typegoose)

**Reply by smolinari** [reply]
I'd also add the suggestion of not breaking away from MongoDB's core ObjectId field. It was built like it is for a number of reasons and using Mongoose/the MongoDB driver takes care of adding the `_id` value for you, simplifying your code somewhat. If you don't use the default `_id` field by changing the type, you are then completely responsible for making sure the ids are added to every new document. So, by changing the id type, you aren't gaining anything and you'll be losing a number advantages (self-creation, ability to sort, getting a timestamp out of the value, etc.)

https://docs.mongodb.com/manual/reference/method/ObjectId/

Also, this won't work. 

```ts
    @prop({default: () => nanoid()})
    _id!: mongoose.Types.ObjectId;
```

Scott

**Reply by vresetnikov** [reply]
Versions:

"mongoose": "^6.2.2",
"@typegoose/typegoose": "^9.6.2",
"typescript": "^4.5.5"

Development environment:
I run the project as an express.js server at localhost on my WebStorm IDE installed on my personal Windows 10 computer. Mongo db is installed also locally and its shell version is 4.2.3, I am testing this with Postman and Robo 3T. I hope this answers your first question.

**Reply by vresetnikov** [reply]
> I'd also add the suggestion of not breaking away from MongoDB's core ObjectId field. It was built like it is for a number of reasons and using Mongoose/the MongoDB driver takes care of adding the `_id` value for you, simplifying your code somewhat. If you don't use the default `_id` field by changing the type, you are then completely responsible for making sure the ids are added to every new document. So, by changing the id type, you aren't gaining anything and you'll be losing a number advantages (self-creation, ability to sort, getting a timestamp out of the value, etc.)
> 
> https://docs.mongodb.com/manual/reference/method/ObjectId/
> 
> Also, this won't work.
> 
> ```ts
>     @prop({default: () => nanoid()})
>     _id!: mongoose.Types.ObjectId;
> ```
> 
> Scott

Yeah Scott, I guess you are right. I was so deeply immersed into solving this problem that I failed to consider whether the juice was worth the squeeze. All this headache could be done away with by simply letting mongodb do its intended job.

**Reply by hasezoey** [reply]
> Development environment:

thanks for providing that information, but with `where and how are you testing this?` i meant "how and where in the code are you using this" (code examples), sorry if this was unclear

or to be more specific: where are you expecting that `id` field to be existing, like in types or at runtime or as a results of a API results?

note: if its as a API Results, then you need to do `document.toString({ virtuals: true })` (or `document.toObject({ virtuals: true })`, because by default mongoose does not include virtuals in the json output

the mentioned options can also be set as schema options

**Reply by vresetnikov** [reply]
@hasezoey Sorry for the late reply, I have given up pursuing this objective as it has proven to be taking away too many built-in advantages for very minor and mostly cosmetic benefits in return.
